File: 01_Data_Structures/Linked_List/LC_023_Merge_k_Sorted_Lists.cpp
/**
 * @file LC_023_Merge_k_Sorted_Lists.cpp
 * @author BalaSeethaRamanjaneyulu
 * @brief Solution for LeetCode 023: Merge k Sorted Lists
 *
 * Problem: You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.
 * Merge all the linked-lists into one sorted linked-list and return it.
 *
 * Platform: LeetCode #023
 * Approach: Divide and Conquer (Merge Sort Style)
 * Time:     O(N log k) where N is total nodes, k is number of lists.
 * Space:    O(log k) due to recursion stack.
 * Link:     https://leetcode.com/problems/merge-k-sorted-lists/
 */

#include <iostream>
#include <vector>

using namespace std;

/**
 * Definition for singly-linked list node.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    /**
     * @brief Merges two sorted linked lists (standard helper).
     */
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        while (l1 && l2) {
            if (l1->val < l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }
        tail->next = l1 ? l1 : l2;
        return dummy.next;
    }

    /**
     * @brief Recursively divides the lists into halves and merges them.
     */
    ListNode* mergeKListsHelper(vector<ListNode*>& lists, int start, int end) {
        if (start > end) return nullptr;
        if (start == end) return lists[start];
        
        int mid = start + (end - start) / 2;
        ListNode* left = mergeKListsHelper(lists, start, mid);
        ListNode* right = mergeKListsHelper(lists, mid + 1, end);
        
        return mergeTwoLists(left, right);
    }

    /**
     * @brief Entry point for merging k lists.
     */
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        return mergeKListsHelper(lists, 0, lists.size() - 1);
    }
};

// ─── Driver & Utilities ──────────────────────────────────────────────────────
void printList(ListNode* head) {
    while (head) {
        cout << head->val << (head->next ? " -> " : "");
        head = head->next;
    }
    cout << endl;
}

ListNode* createList(const vector<int>& nums) {
    if (nums.empty()) return nullptr;
    ListNode* head = new ListNode(nums[0]);
    ListNode* curr = head;
    for (size_t i = 1; i < nums.size(); ++i) {
        curr->next = new ListNode(nums[i]);
        curr = curr->next;
    }
    return head;
}

int main() {
    Solution sol;

    // Test Case: [[1,4,5],[1,3,4],[2,6]] -> [1,1,2,3,4,4,5,6]
    vector<ListNode*> lists;
    lists.push_back(createList({1, 4, 5}));
    lists.push_back(createList({1, 3, 4}));
    lists.push_back(createList({2, 6}));

    cout << "Lists to Merge: " << endl;
    for(auto l : lists) printList(l);

    ListNode* result = sol.mergeKLists(lists);
    cout << "Merged List: "; printList(result);

    return 0;
}

File: 01_Data_Structures/Linked_List/LC_023_Merge_k_Sorted_Lists.java
/**
 * @file LC_023_Merge_k_Sorted_Lists.java
 * @author BalaSeethaRamanjaneyulu
 * @brief Java Solution for LeetCode 023: Merge k Sorted Lists
 */

import java.util.PriorityQueue;

/**
 * Definition for singly-linked list node.
 */
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class LC_023_Merge_k_Sorted_Lists {
    /**
     * Merges k sorted lists using a PriorityQueue (Min-Heap).
     * 1. Add the head of every list to the heap.
     * 2. Repeatedly extract the minimum node and add its 'next' pointer to the heap.
     * 
     * @param lists Array of linked-list heads.
     * @return ListNode The merged sorted list head.
     */
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;

        // Min-heap sorted by node value
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);

        // Add initial heads
        for (ListNode node : lists) {
            if (node != null) {
                pq.add(node);
            }
        }

        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;

        while (!pq.isEmpty()) {
            ListNode minNode = pq.poll();
            tail.next = minNode;
            tail = tail.next;

            if (minNode.next != null) {
                pq.add(minNode.next);
            }
        }

        return dummy.next;
    }

    // ─── Driver & Utilities ──────────────────────────────────────────────────
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + (head.next != null ? " -> " : ""));
            head = head.next;
        }
        System.out.println();
    }

    public static ListNode createList(int[] nums) {
        if (nums.length == 0) return null;
        ListNode head = new ListNode(nums[0]);
        ListNode curr = head;
        for (int i = 1; i < nums.length; i++) {
            curr.next = new ListNode(nums[i]);
            curr = curr.next;
        }
        return head;
    }

    public static void main(String[] args) {
        LC_023_Merge_k_Sorted_Lists sol = new LC_023_Merge_k_Sorted_Lists();

        ListNode[] lists = new ListNode[3];
        lists[0] = createList(new int[]{1, 4, 5});
        lists[1] = createList(new int[]{1, 3, 4});
        lists[2] = createList(new int[]{2, 6});

        ListNode merged = sol.mergeKLists(lists);
        System.out.print("Merged List: "); printList(merged);
    }
}
